worlds/cv64/__init__.py

```diff
@@ -0,0 +1,331 @@
+        self.active_warp_list = []
+        self.reinhardt_stages = True
+        self.carrie_stages = True
+        self.branching_stages = False
+        self.starting_stage = rname.forest_of_silence
+
+        self.total_s1s = 0
+        self.s1s_per_warp = 0
+        self.total_s2s = 0
+        self.required_s2s = 0
+        self.drac_condition = 0
+
+        super().__init__(multiworld, player)
+
+    @classmethod
+    def stage_assert_generate(cls, world) -> None:
+        rom_file = get_base_rom_path()
+        if not os.path.exists(rom_file):
+            raise FileNotFoundError(rom_file)
+
+    def generate_early(self) -> None:
+        # Generate the player's unique authentication
+        self.auth = bytearray(self.multiworld.random.getrandbits(8) for _ in range(16))
+
+        self.total_s1s = self.options.total_special1s.value
+        self.s1s_per_warp = self.options.special1s_per_warp.value
+        self.total_s2s = self.options.total_special2s.value
+        self.required_s2s = int(self.options.percent_special2s_required.value / 100 * self.total_s2s)
+        self.drac_condition = self.options.draculas_condition.value
+
+        # If there are more S1s needed to unlock the whole warp menu than there are S1s in total, drop S1s per warp to
+        # something manageable.
+        while self.s1s_per_warp * 7 > self.total_s1s:
+            self.s1s_per_warp -= 1
+
+        # Enable/disable character stages and branching paths accordingly
+        if self.options.character_stages.value == self.options.character_stages.option_reinhardt_only:
+            self.carrie_stages = False
+        elif self.options.character_stages.value == self.options.character_stages.option_carrie_only:
+            self.reinhardt_stages = False
+        elif self.options.character_stages.value == self.options.character_stages.option_both:
+            self.branching_stages = True
+
+        self.active_stage_exits = get_normal_stage_exits(self)
+
+        stage_1_blacklist = []
+
+        # Prevent Clock Tower from being Stage 1 if more than 4 S1s are needed to warp out of it.
+        if self.s1s_per_warp > 4 and not self.options.multi_hit_breakables.value:
+            stage_1_blacklist.append(rname.clock_tower)
+
+        # Shuffle the stages if the option is on.
+        if self.options.stage_shuffle:
+            self.active_stage_exits, self.starting_stage, self.active_stage_list = \
+                shuffle_stages(self, stage_1_blacklist)
+        else:
+            self.active_stage_list = [stage for stage in vanilla_stage_order if stage in self.active_stage_exits]
+
+        # Create a list of warps from the active stage list. They are in a random order by default and will never
+        # include the starting stage.
+        self.active_warp_list = generate_warps(self)
+
+    def create_regions(self) -> None:
+        # Create the Menu region.
+        active_regions = {"Menu": Region("Menu", self.player, self.multiworld)}
```

Comment:
That was written before the world convenience getters were added, and I believe at the time I had concerns with calling the multiworld getters many times over being very slow in larger multis.

Reply:
the world getters are new enough that's fine but the multiworld getters were only problematic before https://github.com/ArchipelagoMW/Archipelago/commit/3e0d1d4e1c6f34605aba1e36480f341e812fe26c because using them during region creation would cause recaching, which is no longer the case.